### ISSUE TITLE
Add Stored Knowledge

### ISSUE THREAD

**Issue by VincenzoFerme** [benchflow/benchflow#381]
@jeppe-style write down the issues when you are on this, according to the discussion we had that is summarised in the attached paper.

![img_20170530_124102](https://cloud.githubusercontent.com/assets/7163641/26579745/630cb61c-4535-11e7-92f1-5553a1d23cdb.jpg)

**Comment by jeppe-style** [comment]
We want to reuse the knowledge already gathered from previous experiment runs. It is done in the `add stored knowledge` state in the test-manager. 

The basic idea is to compute a hash of the experiment bundle and store that in the DB after the experiment has been executed in a reusable state:
- TERMINATED - COMPLETED: reuse metrics, result
- TERMINATED - FAILURE: reuse result
- TERMINATED - ABORTED: no reuse possible
- TERMINATED - ERROR: reuse result

When we get a new experiment we can hash that experiment bundle and look for a match in the DB. If there is a match we do a detail comparison to ensure they are really the same. Since not all data is relevant (e.g. name of experiment) we first remove this data before computing the hash.

To compute the hash the process looks like this:
1. Get the experiment bundle from minio
2. Remove data that is not relevant from the experiment definition: `name`, `description`
4. TODO - we need to ensure that spaces, order of elements (sort maps) are not influencing the hash so that it cannot be easily manipulated (e.g. add a space or change order of some elements)
3. Zip the bundle to a `ZipInputStream`
4. Compute the MD5 hash using [DigestUtils](http://commons.apache.org/proper/commons-codec/apidocs/org/apache/commons/codec/digest/DigestUtils.html) (@VincenzoFerme I think MD5 should be fine although I didn't find any reliable resource to motivate it)

To add stored knowledge:
1. For each experiment in the exploration space:
   1. Compute the hash as describe above
   2. Search in the DB for matching hash
   3. IF match found: compare the contents of the experiment bundles (convert to string and compare strings)
   4. IF match by content: save the experiment model in the DB and copy the status from the matched experiment

We also need to add an optional value in the DSL so that the expert user can turn the feature off and force the execution even if a match was found. Covered by https://github.com/benchflow/benchflow/issues/444

**Comment by VincenzoFerme** [comment]
```
The basic idea is to compute a hash of the experiment bundle and store that in the DB
```
Only when the experiment has successfully completed the execution in a state that marks the results as reusable, see: https://github.com/benchflow/benchflow/issues/381#issue-232211893. It does not make sense you save the hash of something you can't reuse.

**Comment by VincenzoFerme** [comment]
For future improvements: since we might have similar experiments in the queue that are not yet executed, we can also check for them and if they match the hash we wait for the execution. Moreover we should also check for already existing knowledge just before executing something, because in the meantime new experiments might have been executed.

**Comment by jeppe-style** [comment]
I updated the comment according to https://github.com/benchflow/benchflow/issues/381#issue-232211893

**Comment by VincenzoFerme** [comment]
Ok, so for:

- TERMINATED - ABORTED: no reuse possible

You do not store it, and then for point 1, iii in https://github.com/benchflow/benchflow/issues/381#issuecomment-306450369 you need to check the actual state to understand what you can actually reuse.

**Comment by jeppe-style** [comment]
I had forgotten to remove to check that the state is TERMINATED.

Since we only store the hash after the experiment has been executed in a reusable state I think it is sufficient to copy the data associated from the reusable experiment since it must contain the relevant information. It can then be handled like as if it would have been executed and there is not need to check the state besides what would happen if it would have been executed in the normal way.

**Comment by VincenzoFerme** [comment]
Also for the ABORTED case?

**Comment by jeppe-style** [comment]
The ABORTED case is identified as not reusable and as such we would not compute the hash on that experiment bundle.

**Comment by VincenzoFerme** [comment]
@jeppe-style - Make Reuse Stored Knowledge more resilient to changes to the DSL specification as discussed in the last meeting. Since you now have `ExplorationModel` holding part of the information to determine if you can reuse something, you might think of having another class to store the remaining information you need for reuse, and then optimise that with hashing, so that you control the structure of the data, and you are sure that you can extend it over time, taking into account the fact the data can grow and change for the checks you perform.

**Comment by jeppe-style** [comment]
@VincenzoFerme I don't see the benefits compared to hashing all the files in the experiment bundle.

**Comment by VincenzoFerme** [comment]
In the YAML you could:

- change name of fields often
- have a different order of the elements in different submissions

So it is fine to hash part of the bundle (e.g., the model folder for which I would only store the hash), but for the remaining files (using the BenchFlow DSL), that would not be safe.

**Comment by jeppe-style** [comment]
- _change name of fields often_ => I don't think we should promote changing the names of the fields. A new version of the DSL should be released for that.
- _have a different order of the elements in different submissions_ => this we need to take care of during serialization and should therefore be deterministic

**Comment by VincenzoFerme** [comment]
I agree, but if you use the hash of two different version of the DSL that describe the same test, you do not catch the reuse possibility by using what you propose.

**Comment by VincenzoFerme** [comment]
So the outcome of the discussion above is that:

- we need to make the reuse base functionality less tight to the DSL we have, so that we can increase spotting reuse opportunities.

My proposal is sketched in the following image:

![img_20170626_174827](https://user-images.githubusercontent.com/7163641/27547983-d6cc908c-5a97-11e7-91ba-8060965e6c15.jpg)

Before implementing it we can discuss the idea and cleanup this issue.
